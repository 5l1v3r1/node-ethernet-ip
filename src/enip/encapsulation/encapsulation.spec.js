const encapsulation = require("./index");
const net = require("net");

describe("Encapsulation", () => {
    describe("Command Validator", () => {
        const { validateCommand } = encapsulation;
        const {
            RegisterSession,
            UnregisterSession,
            SendRRData,
            SendUnitData
        } = encapsulation.commands;

        it("Rejects Invalid Commands", () => {
            expect(validateCommand(0x99)).toBeFalsy();
            expect(validateCommand("hello")).toBeFalsy();
            expect(validateCommand(0x02)).toBeFalsy();
        });

        it("Accepts Proper Commands", () => {
            expect(validateCommand(0x66)).toBeTruthy();
            expect(validateCommand(102)).toBeTruthy();
            expect(validateCommand(RegisterSession)).toBeTruthy();
            expect(validateCommand(UnregisterSession)).toBeTruthy();
            expect(validateCommand(SendRRData)).toBeTruthy();
            expect(validateCommand(SendUnitData)).toBeTruthy();
        });
    });

    describe("Status Parser", () => {
        const { parseStatus } = encapsulation;

        it("Rejects Non-Number Inputs", () => {
            expect(() => parseStatus("test")).toThrow();
            expect(() => parseStatus(null)).toThrow();
            expect(() => parseStatus(undefined)).toThrow();
        });

        it("Returns Proper Human Readable String", () => {
            expect(parseStatus(0)).toEqual("SUCCESS");
            expect(parseStatus(0x01)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(1)).toEqual(expect.stringContaining("FAIL"));
            expect(parseStatus(0x45)).toEqual(expect.stringContaining("FAIL"));
        });
    });

    describe("Header Building Utility", () => {
        const { header: { build }, commands: { RegisterSession } } = encapsulation;

        it("Builds Correct Encapsulation Buffer", () => {
            const snap = build(RegisterSession, 0x00, [0x01, 0x00, 0x00, 0x00]);
            expect(snap).toMatchSnapshot();
        });
    });

    describe("Header Parsing Utility", () => {
        const { header: { parse, build }, commands: { SendRRData } } = encapsulation;

        it("Builds Correct Encapsulation Buffer", () => {
            const data = build(SendRRData, 98705, [0x01, 0x00, 0x00, 0x00]);
            const snap = parse(data);

            expect(snap).toMatchSnapshot();
        });
    });

    describe("Test Encapsulation Generator Functions", () => {
        const { registerSession, unregisterSession, sendRRData, sendUnitData } = encapsulation;

        it("Register Session Returns Correct Encapsulation String", () => {
            const data = registerSession();

            expect(data).toMatchSnapshot();
        });

        it("Unregister Session Returns Correct Encapsulation String", () => {
            const data = unregisterSession(98705);
            
            expect(data).toMatchSnapshot();
        });

        it("SendRRData Returns Correct Encapsulation String", () => {
            const data = sendRRData(98705, Buffer.from('hello world'));

            expect(data).toMatchSnapshot();
        });

        it("SendUnitData Returns Correct Encapsulation String", () => {
            const data = sendUnitData(98705, Buffer.from("hello world"));

            expect(data).toMatchSnapshot();
        });
    });
});
