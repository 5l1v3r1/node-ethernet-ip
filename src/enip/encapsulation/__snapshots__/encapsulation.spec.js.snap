// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encapsulation Header Building Utility Builds Correct Encapsulation Buffer 1`] = `
Object {
  "data": Array [
    101,
    0,
    4,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Encapsulation Header Parsing Utility Builds Correct Encapsulation Buffer 1`] = `
Object {
  "command": "SendRRData",
  "commandCode": 111,
  "data": Object {
    "data": Array [
      1,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  "length": 4,
  "options": 0,
  "session": 98705,
  "status": "SUCCESS",
  "statusCode": 0,
}
`;
